### ISSUE TITLE
Some early feedback

### ISSUE THREAD

**Issue by jedisct1** [samuel-lucas6/draft-lucas-balloon-hashing#27]
Hi Samuel,

In case this is useful, here's some random early feedback.

I wrote an implementation based on that document (https://github.com/jedisct1/zig-bkdf), and the test vectors didn't pass. But I may have overlooked or screwed up something.
Is there a reference implementation closely matching the document somewhere?

Anyway. Here are some observations.

---

> ByteArray(l): the creation of a new byte array with length

Should that array be initialized with zeros?

In the `BKDF` steps, `hash` is not initialized:

```
hash = ByteArray(HASH_LEN)
foreach output in outputs
    for i = 0 to output.Length - 1
        hash[i] = hash[i] ^ output[I]
```

So, either `ByteArray` should set everything to zero, or `hash` should be set to `output[0]` and then `output[1..]` should be mixed.

---

I found the upper bound of the `for` loops ranges to be very confusing, as the document uses inclusive ranges, whereas programming languages generally don't. 

---

```
if pepper.Length == 0
    key = ByteArray(0)
else
    key = pepper
key = ZeroPad(key, KEY_LEN)
```

Why the special case when the length is `0`?

---

```
key = PRF(key, password || salt || personalization || associatedData || LE32(pepper.Length) || LE32(password.Length) || LE32(salt.Length) || LE32(personalization.Length) || LE32(associatedData.Length))
```

Here, the `key` input parameter and the `key` output parameters represent different things, and have different lengths. Maybe use different names?

---

```
spaceCost = 2 ** spaceCost
```

The same variable uses different units at different places, and this is a bit confusing.

Maybe rename the one representing the log2 of the space cost `spaceCostLog`?

---

```
LE32(spaceCost)
```

`spaceCost` must fit in 32-bits. But `MAX_SPACECOST` is `32`, and `1<<32` would overflow.

Maybe absorb `LE32(spaceCostLog)` instead of `LE32(spaceCost)`? This will also lift the 2^32 limitation.

---

```
reps = (spaceCost * timeCost * 3) / (HASH_LEN / 4)
```

This assumes that HASH_LEN is a multiple of 4. Which is the case for virtually all hash functions, but is still not universally true.

---

BKDF only needs a PRF. It doesn't have to be a hash-based PRF. So, maybe it would be better to use something like `PRF_OUT_LEN` rather than `HASH_LEN`?

---

> KEY_LEN: the length of the PRF key in bytes. This MUST be the length of whichever is shortest: the hash function block size or the maximum PRF key size. For example, 128 bytes for HMAC-SHA512 and SHA-512 (the block size) but 64 bytes for BLAKE2b and 32 bytes for BLAKE3 (the maximum key parameter size)

The test vectors use SHA256, so maybe the key length for that one should be included?

---

Test Vector 3:

```
spaceCost: 1024
```

How could that test vector pass if this represents 2^1024 hash outputs?

**Comment by jedisct1** [comment]
> It is RECOMMENDED to use a collision-resistant hash function with a larger output length that is fast in software but relatively slow in hardware, such as BLAKE2b-512. As another example, SHA-512 is preferable to SHA-256.

I'm not sure about the last part. Most modern CPUs have SHA-256 hardware acceleration.

However, SHA-512 acceleration on Intel CPUs is very new (only available on Lunar Lake CPUs released 3 months ago).

**Comment by jedisct1** [comment]
> HMAC [[RFC2104](https://samuellucas.com/draft-lucas-bkdf/draft-lucas-bkdf.html#RFC2104)] is also NOT RECOMMENDED because it increases the number of calls to the chosen hash function.

~~The rest of the document recommends using HMAC to build a keyed hash function.~~

My bad, I overlooked this in the definitions:

> if there is no key parameter (e.g. SHA-512 [[RFC6234](https://samuellucas.com/draft-lucas-bkdf/draft-lucas-bkdf.html#RFC6234)]), you MUST perform prefix MAC and pad the key with zeros to the block size (1024 bits for SHA-512)

Maybe the definition of the `PRF()` could benefit from a dedicated section, that clearly explains how it's defined for hash functions, MACs and XOFs.

**Comment by jedisct1** [comment]
> A smaller output length worsens the performance and memory-hardness.

Why does it worsen memory hardness ?

**Comment by samuel-lucas6** [comment]
> I wrote an implementation based on that document (https://github.com/jedisct1/zig-bkdf), and the test vectors didn't pass. But I may have overlooked or screwed up something.

Thanks Frank, that's awesome. I need to update the test vectors as they were for the Rust Crypto Balloon implementation. I was waiting until I'd finished tampering with the design.

> Is there a reference implementation closely matching the document somewhere?

Not yet. This is something I've been meaning to do.

> Should that array be initialized with zeros?

I should probably do that and then use `ZeroPad()` less.

> So, either ByteArray should set everything to zero, or hash should be set to output[0] and then output[1..] should be mixed.

Yes, I expect the latter would be clearer.

> I found the upper bound of the for loops ranges to be very confusing, as the document uses inclusive ranges, whereas programming languages generally don't.

I think this is a consequence of following the paper. That sort of thing can definitely be changed.

> Why the special case when the length is 0?

That was just to make it clear that it's empty, but you're right that that should probably be changed.

> Here, the key input parameter and the key output parameters represent different things, and have different lengths. Maybe use different names?

Yeah, I will look into that. I think I was trying to have as few variables as possible, and the key gets padded again afterwards.

> The same variable uses different units at different places, and this is a bit confusing.
> 
> Maybe rename the one representing the log2 of the space cost spaceCostLog?

I agree that's not done well. I think this was for the same reason as above but is something I've been planning to change if I can switch from blocks to KiB/MiB.

> spaceCost must fit in 32-bits. But MAX_SPACECOST is 32, and 1<<32 would overflow.
> 
> Maybe absorb LE32(spaceCostLog) instead of LE32(spaceCost)? This will also lift the 2^32 limitation.

Yep, I've been thinking about this too. I thought `MIN_SPACECOST` was 1, leaving room for a weird wrap around, but it isn't for some reason.

> This assumes that HASH_LEN is a multiple of 4. Which is the case for virtually all hash functions, but is still not universally true.

Do you mean XOFs?

> BKDF only needs a PRF. It doesn't have to be a hash-based PRF. So, maybe it would be better to use something like PRF_OUT_LEN rather than HASH_LEN?

I've written the document on the basis that hash functions are being used at the moment, but this is something that could be changed. I agree the PRF vs hash naming/discussion is confusing and should be consistent.

> The test vectors use SHA256, so maybe the key length for that one should be included?

That's a good point.

> How could that test vector pass if this represents 2^1024 hash outputs?

The test vectors desperately need to be updated as they're for Rust Crypto's Balloon implementation. I either need to do a reference implementation that gets updated alongside the draft and do a few test vectors for now, or I could wait until the whole thing has been finalised and do all the test vectors in one go.

> I'm not sure about the last part. Most modern CPUs have SHA-256 hardware acceleration.
> 
> However, SHA-512 acceleration on Intel CPUs is very new (only available on Lunar Lake CPUs released 3 months ago).

Yeah, this comment is really about the block size (see below) but can be tweaked.

> Maybe the definition of the PRF() could benefit from a dedicated section, that clearly explains how it's defined for hash functions, MACs and XOFs.

Good idea.

> Why does it worsen memory hardness ?

I think this came from me reading [this](https://ieeexplore.ieee.org/document/7961977) paper ages ago:

> The attacks of [AB16] perform particularly well when the block size is small. In particular, the (potentially) smaller blocksize of BH is a disadvantage (in comparison with Argon2i-B) as it means that we need to select a higher value of σ to achieve the same memory usage and the attack quality of [AB16] increase rapidly with n. This would make BH less resistant to the attacks than Argon2i-B.

Whether I've interpreted the paper correctly is another matter. Password hashing papers aren't the most readable for me.

**Comment by solardiz** [comment]
I just learned of this repo today. I was hoping Balloon would just stay academic-only. It's much weaker than, say, yescrypt for practical purposes. But I can see that there may be demand for a single-primitive algorithm like this. We just do not have a standards-compliant single primitive that would make such algorithm as strong, unfortunately - which is why all of scrypt, yescrypt, and Argon2 came up with their own primitives and block functions.

As I recall the paper (from seeing it years ago), a focus of Balloon was greater TMTO resistance compared to Argon2, however:
1. The improvement in TMTO resistance is (more than) negated by lower total memory usage per time.
2. I currently think that TMTO resistance is much better achieved by making the block function stateful as done in yescrypt 1.0 (the state size is such that it fits L1 cache, and so doesn't take away from memory bandwidth), and not so much by accessing blocks in certain carefully chosen/analyzed patterns (although this is also important) nor increasing the number of iterations over memory. Due to this, I expect that yescrypt already has more than sufficient TMTO resistance even at t=0.

So I really see little point in pursuing Balloon further. But if there's demand anyway and standardization and adoption is going to happen anyway, I don't mind contributing to clarifying things here, so that the trade-offs are understood.

I also commend on the effort anyhow - this repo/draft looks well maintained - kudos!

> > A smaller output length worsens the performance and memory-hardness.
> 
> Why does it worsen memory hardness ?

Small output length means lower memory filling/processing performance. Lower performance translates to lower memory bandwidth usage and to less memory used during the time allotted. Small block size with random access also means more waiting on memory latency, which even further reduces the amount of memory used during the time allotted, although SHA-512 is so slow on its own that this additional effect is small.

Then the attacker introduces extra parallelism (more threads, SIMD, GPU, ASIC, etc.) and achieves higher cumulative memory bandwidth yet is able to fit the many more instances in a moderate amount of memory (since not that much memory per instance was used defensively - perhaps an order of magnitude less than state-of-the-art memory-hard KDFs would).

I recently played with attacking an scrypt-inspired memory-hard single-primitive KDF based solely on SHA-512 that's used by Armory wallet. On AVX-512, it's basically ~8x faster to attack (compared to 64-bit scalar optimized code that is the best that could be used defensively), as implemented here:

https://github.com/openwall/john/blob/08021aa26b6010e9baf6865956b7326597f23e01/src/armory_fmt_plug.c#L283

Even with AVX-512's ~8x attack speed boost, and all cores and hardware threads in use, it's still way below a typical CPU+RAM system's available memory bandwidth (but finally a significant fraction of the bandwidth, IIRC something like 40% total).

One way around that could be to also embrace SIMD defensively (design the KDF's configurable parallelism such that it's available not only at thread, but also at SIMD level), but this defeats the purpose of simplicity vs. scrypt, etc. Even if we come up with design that's not too complex, implementations that actually take advantage of it would be.

And then if someone configures, say, 32x parallelism to allow 4 threads x 8-lane AVX-512 defensively, but an actual system is built for, say, base x86_64 for portability, so up to 128-bit SSE2, then it'd be even making things worse - giving a free gift of 4x parallelism to attacker only. OTOH, this is also the case with Argon2's commonly excessive built-in parallelism.

Disclosure: I designed yescrypt.

**Comment by solardiz** [comment]
Another way to look at this is that when instantiated with SHA-512, there appears to be too little advantage of BKDF vs. PBKDF2 (and SHA-256 probably even worse, although SHA-NI should skew this a little bit).

The draft has the below, I guess data from @sc00bz?

```
To cause an attacker to get < 10 kH/s on an RTX 4080 SUPER GPU, the parameters must be at least one of the following when using SHA-256, SHA-512, HMAC-SHA256, or HMAC-SHA512 as the collision-resistant PRF:

- SHA-256: `m=256 KiB, t=48` or `m=512 KiB, t=24` or `m=1 MiB, t=12` or `m=2 MiB, t=6`
- SHA-512: `m=256 KiB, t=34` or `m=512 KiB, t=17` or `m=1 MiB, t=9` or `m=2 MiB, t=4`
```

If I'm reading the pseudo-code and doing my math right, `m=2 MiB, t=4` here corresponds to 10 MiB output by SHA-512 and 42 MiB written+read to+from memory. 10 MiB output means `10*2^20/64 = 163840` SHA-512's computed. Incidentally, that figure is typical for SHA-512 iteration count seen in Bitcoin Core wallet.dat (which automatically tunes for acceptable defensive running time), so corresponds to the KDF rather than password hashing use case, meaning there isn't a lot of room to increase it further defensively.

At 10k/s (unfortunately, no exact speed is given here), the speed is `10*2^20/64*10000/10^9 = 1.64` billion SHA-512/s, whereas 42 MiB means ~440 GB/s, which is a respectable 60% of RTX 4080 SUPER's specified 736 GB/s memory bandwidth.

For comparison, https://hashcat.net/forum/thread-11837.html has some other algorithms benchmarked on RTX 4080 SUPER. A relevant one is:

```
* Hash-Mode 7100 (macOS v10.8+ (PBKDF2-SHA512)) [Iterations: 1023]
------------------------------------------------------------------

Speed.#1.........:  1825.3 kH/s (69.97ms) @ Accel:16 Loops:255 Thr:512 Vec:1
```

That's `1023*1825.3*1000*2/10^9 = 3.73` billion SHA-512/s when used in PBKDF2-HMAC-SHA512. That's ~2.27x faster.

So it looks like the effect of BKDF is maybe a ~2.5x slowdown of GPU attacks (considering that actual speed was `< 10 kH/s` and not exactly 10k), but then I doubt the implementation is fully optimized yet and there may also be some slowdown (resulting from the memory accesses) on CPU. So the improvement in terms of attack/defense over PBKDF2 is probably less than 2x?

2 MiB is probably not the maximum that would reasonably be used, but it's so low that I expect even several times larger sizes would still allow the GPU to achieve full speed without being limited by its global memory size yet.

2 MiB is also very low against ASICs, yet non-negligible. For Monero original CryptoNight mining, which used exactly 2 MiB, ASICs achieved a speedup of roughly 100x over CPUs and GPUs per-Watt:

https://github.com/openwall/john/issues/5668#issuecomment-2693130816

CPUs and GPUs were roughly same speed for that algorithm because it heavily used AES-NI, which GPUs had to emulate. For BKDF with SHA-512, I think the ASIC advantage will be > 100x because BKDF's memory bandwidth usage is many times lower.

**Comment by samuel-lucas6** [comment]
Hi @solardiz, nice to see you here, and thank you for your comments. Sorry for the slow reply; I've been a bit busy IRL recently.

I agree with your points. For context, I started this project because a) I liked the simplicity of Balloon and b) it felt like there was potential getting it adopted due to the simplicity (think HMAC, HKDF, ChaCha20, sponge/duplex, etc) and use of NIST compliant functions (aka a PBKDF3 candidate). However, I ran into the following issues:

1. The authors of the Balloon paper made a mistake defining the function in a theoretical rather than practical sense. This has caused interoperability issues and harmed interest in the algorithm (from where I'm sitting). An implementer/specification writer has to semi-design the algorithm.
2. As you point out, the simplicity makes it very slow and suboptimal from a security standpoint, making it a questionable improvement over PBKDF2 compared to other algorithms. (Those figures are from Sc00bz but not from him implementing Balloon/BKDF from memory, just basing it on PBKDF2 or something.)
3. Having any generic design like this is difficult because of the differences between (keyed) hash functions. For example, unkeyed BLAKE2/BLAKE3 vs keyed, SHA-2 vs HMAC-SHA2, SHA-3 vs SHAKE vs KMAC vs KangarooTwelve. I think this is partly why HKDF hasn't really been replaced. These generic constructions are near impossible to do cleanly now.
4. NIST suggested they want to approve Argon2 in an email to me when I contacted them. They didn't seem interested in Balloon/BKDF, even though Balloon has been mentioned in a [NIST document](https://csrc.nist.gov/pubs/sp/800/63/b/upd2/final) before (despite it having no specification/interoperable design).
5. There's little to no academic or community interest in finding better password hashing algorithms now. There aren't many new papers, and most people are content with Argon2. I think people are either unaware of Argon2's limitations/design flaws, consider it good enough, or are more concerned with other topics like PQC or something they better understand. And Argon2 has taken a while to get deployed, so there's no rush to do that again.
6. There are lots of open questions/unclear answers to questions, like memory-hardness vs cache-hardness, data-dependent vs independent vs hybrid, etc. Without the above, these questions won't be answered properly, and a designer is on their own.
7. I'm the wrong person for the job in that I was still in school (not uni) when the PHC happened and have no background in password cracking/hardware. I haven't gone through much of the PHC, password hashing is barely covered in cryptography books for some reason (even by JP), and the literature is among the least accessible on the symmetric crypto side. I'm also a strong believer in collaboration from a security/best design perspective, but that doesn't work in reality unless you know people or there's sufficient interest in what you're doing.

So, I'm not really sure where to go from here. I came up with [Finnick](https://github.com/samuel-lucas6/Finnick) as a way to speed up Balloon. The idea being to use a faster function (e.g., something with AES-NI) whilst retaining the simplicity. I also came across [Passcrypt](https://github.com/rmlibre/aiootp?tab=readme-ov-file#passcrypt--table-of-contents). However, multiple of the above points still stand, and I've been focusing on other projects lately, like [ChaCha20-Poly1305-SIV](https://github.com/C2SP/C2SP/pull/130/commits/ceb8871280392d4f439dc03b0c80c6ad849073e8).

My primary motivation for projects is to learn/practice/help other people, with interest/adoption being a bonus. I don't expect to be taken very seriously given I lack experience and don't work in this field. It's a hobby because it's interesting.

**Comment by solardiz** [comment]
Hi @samuel-lucas6. Thank you very much for your comments! I find your perspective relatable and reasonable.

I am curious what you think "Argon2's limitations/design flaws" are. IIRC, the authors of the Balloon paper thought it was unmet TMTO resistance claims, but to me that's purely of academic value (just revise the claims).

I agree there's room for much further research. For example, "memory-hardness vs cache-hardness" can be answered with "both" (as done in yescrypt and TwoCats), but no one has yet looked into data-independent cache-hardness.

Thank you for the Finnick and Passcrypt references. I don't recall seeing them before. Are there any benchmarks of these?

I agree AES-NI (and equivalents on other archs) may be a good way to achieve sufficient memory filling/processing speed. It also works against GPUs on its own, so reduces the need for cache-hardness. A drawback is that the fallback to software AES may expose the scheme to cache timing attacks. Curiously, the initial Argon submission (before Argon2) was based on AES, but unfortunately it was awfully slow (I don't recall exactly why).

Oh, I also had a question to you: what does "Mode of operation" mean in the comparison table here in the BKDF spec draft?

**Comment by samuel-lucas6** [comment]
Thank you @solardiz.

> I am curious what you think "Argon2's limitations/design flaws" are. IIRC, the authors of the Balloon paper thought it was unmet TMTO resistance claims, but to me that's purely of academic value (just revise the claims).

One thing I was thinking of is how it's supposedly worse than bcrypt at shorter runtimes. However, I think the only evidence I've seen of this is a [screenshot](https://infosec.exchange/@epixoip/110912922574721750) from Jeremi Gosney that lacks much context. Anyhow, this made it onto the [bcrypt Wikipedia page](https://en.wikipedia.org/wiki/Bcrypt#Comparison_to_other_password_hashing_algorithms) and is something several people have said.

Another example is [this](https://eprint.iacr.org/2018/944) paper, which claims the Argon2 round function can be made more sequential. I spoke to Jeremiah Blocki and Joël Alwen briefly, and they were talking about how the Argon2 'edge structure' can be improved. They were also interested in [trapdoor memory-hard functions](https://eprint.iacr.org/2024/312), although they sound worse than using a pepper to me.

Then there's been some criticism of Argon2's approach to parallelism, like you/[Sc00bz](https://github.com/samuel-lucas6/draft-lucas-bkdf/issues/17) have touched on, which is also what complicates the algorithm. I find the way they use BLAKE2b initially and to generate output a bit non-standard as well, like the position of the key and discarding half the output when doing multiple blocks.

> I agree there's room for much further research. For example, "memory-hardness vs cache-hardness" can be answered with "both" (as done in yescrypt and TwoCats), but no one has yet looked into data-independent cache-hardness.

I need to look into those at some point.

> Thank you for the Finnick and Passcrypt references. I don't recall seeing them before. Are there any benchmarks of these?

I haven't done a proper implementation of Finnick, and other people will be able to do a more efficient implementation than me. I did a proof of concept with libsodium's [AEGIS](https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-aegis-aead) API, which was faster than Balloon/BKDF. However, that API is for an AEAD rather than a stream cipher, and it requires copying blocks into one associated data parameter rather than just accessing and absorbing the blocks into the state. I expect it's still slower than Argon2 though and currently requires two primitives, although there may be ways around that. There's now [HiAE](https://eprint.iacr.org/2025/377) as an alternative to AEGIS.

I don't think there are benchmarks for Passcrypt either. It sounds interesting but is based on SHAKE128, which is probably not ideal against hardware.

> I agree AES-NI (and equivalents on other archs) may be a good way to achieve sufficient memory filling/processing speed. It also works against GPUs on its own, so reduces the need for cache-hardness. A drawback is that the fallback to software AES may expose the scheme to cache timing attacks. Curiously, the initial Argon submission (before Argon2) was based on AES, but unfortunately it was awfully slow (I don't recall exactly why).

Yes, I've been a bit confused about whether AES-NI is a good idea or not. I noticed Argon used AES but didn't find out why they switched, and I've seen [papers](https://ieeexplore.ieee.org/document/9422754) that imply AES can be fast on GPUs.

> Oh, I also had a question to you: what does "Mode of operation" mean in the comparison table here in the BKDF spec draft?

I took that phrasing from the Balloon paper. I mean whether the algorithm can be instantiated with a different function trivially. It overlaps with `Single primitive` really.

**Comment by solardiz** [comment]
Thank you @samuel-lucas6 for referencing so much interesting and relevant material.

> One thing I was thinking of is how it's supposedly worse than bcrypt at shorter runtimes. However, I think the only evidence I've seen of this is a [screenshot](https://infosec.exchange/@epixoip/110912922574721750) from Jeremi Gosney that lacks much context.

I didn't compare Argon2 against bcrypt lately, but one metric for memory-hard hashes I'm considering is the break-even point for cracking on GPU vs. on CPU (where the latter should for an ideal KDF/hash not be much faster than defensive use on CPU). That is, the parameter values where further increase in memory hardness makes the GPU slower than a "comparable" CPU (e.g. contemporary to the GPU and with similar power consumption). For Argon2 as we currently have it in JtR, this point is in the hundreds of megabytes, which is rather high. Unfortunately, we do not yet have an optimized GPU implementation of yescrypt 1.0, but from mining benchmarks of yescrypt 0.5 I expect its break-even point is way lower.

@epixoip's results in that screenshot show a related phenomenon. bcrypt on whatever device slows down linearly with 2^cost, whereas Argon2 at low m is quite GPU-friendly and it slows down linearly for low m, but asymptotically as m^2. So I am not surprised the break-even point vs. bcrypt on a certain GPU (apparently, AMD RX 7900 XTX per a comment further in the thread) may also be in the hundreds of megabytes, although some other attacker would choose a bcrypt-friendlier GPU instead (perhaps recent NVIDIA). I note these benchmarks are from August 2023, whereas we got Argon2 OpenCL code into JtR and optimized it further starting December 2023. More context would indeed be desirable.

> Another example is [this](https://eprint.iacr.org/2018/944) paper, which claims the Argon2 round function can be made more sequential.

That's one lengthy paper. Per the abstract, their focus appears to be once again on the academic TMTO resistance for data-independent access. I couldn't quickly find where they say "the Argon2 round function can be made more sequential".

> I spoke to Jeremiah Blocki and Joël Alwen briefly, and they were talking about how the Argon2 'edge structure' can be improved. They were also interested in [trapdoor memory-hard functions](https://eprint.iacr.org/2024/312), although they sound worse than using a pepper to me.

To me, trapdoor memory-hard functions sound like they're for some different use cases, and I wonder what those would be.

As to pepper, to me it's clear (and was at the time of PHC) that we need hash encryption instead of pepper (which is why yescrypt only has hash encryption). It's as good as pepper, but it also allows for key change (if potentially compromised). Doing it without dependency on an extra primitive is a bit tricky. yescrypt has a custom Feistel cipher based on SHA-256. It's not very complex, but I understand that when you optimize for simplicity, you implement a pepper instead.

> Then there's been some criticism of Argon2's approach to parallelism, like you/https://github.com/samuel-lucas6/draft-lucas-bkdf/issues/17 have touched on, which is also what complicates the algorithm.

Looks like you misunderstood me. I didn't comment on Argon2's explicit/configurable parallelism at all. I commented on the lower-level parallelism hard-coded in its block function. This is what allows efficient use of GPU compute resources even when computing relatively few distinct Argon2 hashes at a time. With Argon2, we have 32 concurrent 32-bit computations for one hash computation with p=1, e.g. see how 480 concurrent hashes translate to 15360 OpenCL work items here: https://www.openwall.com/presentations/OffensiveCon2024-Password-Cracking/slide-42.html

This isn't fully avoidable in KDF design when trying to have enough parallelism for modern CPUs, but 32x per thread is excessive considering that concurrent computation of many Argon2 hashes bumps into memory bandwidth so this parallelism is wasted in defensive use cases where Argon2 settings are limited by target server request rate capacity. In a non-memory-hard KDF, 32x or even more would be reasonable, but here we have a tradeoff.

> I've been a bit confused about whether AES-NI is a good idea or not. I noticed Argon used AES but didn't find out why they switched, and I've seen [papers](https://ieeexplore.ieee.org/document/9422754) that imply AES can be fast on GPUs.

Sure AES can be fast on GPUs, but AES-NI makes it faster than other primitives on CPUs, so the GPU advantage is reduced. As to this specific paper, we actually looked into it and made code changes in JtR accordingly:

https://github.com/openwall/john/issues/5594
https://github.com/openwall/john/pull/5613

I just ran some benchmarks on the same GPU vs. CPUs as were used for the Argon2 benchmark I referred to above (where the GPU was 4x faster than the CPUs). For algorithms involving iterated AES (not memory hard) the GPU (with optimized AES OpenCL code as above) is only 2x to 3x faster than the CPUs (using AES-NI). But these are rather old CPUs and our AES-NI code is currently not interleaved (so performance is limited by instruction latency rather than throughput). With modern AES-NI or better yet VAES (4x parallel AES in 512-bit vectors) and some interleaving (although including the parallelism to allow this has the tradeoff), modern many-core CPUs can probably deliver similar AES speeds as modern GPUs.

**Comment by solardiz** [comment]
> I just ran some benchmarks on the same GPU vs. CPUs as were used for the Argon2 benchmark I referred to above (where the GPU was 4x faster than the CPUs). For algorithms involving iterated AES (not memory hard) the GPU (with optimized AES OpenCL code as above) is only 2x to 3x faster than the CPUs (using AES-NI).

For comparison, on the same GPU vs. CPUs, PBKDF2-HMAC-SHA512 is 13.6x faster to crack on the GPU than on the CPUs, and PBKDF2-HMAC-SHA256 18.6x faster. This is against CPU _cracking_ speeds on 128-bit AVX, so the figures against typical defensive speeds would be 2x and 4x higher than these (~27x and ~74x). So Argon2 _does_ have advantage over PBKDF2, but simple iterated AES algorithms have even greater advantage as long as the defender uses AES-NI (could be on par with PBKDF2 otherwise). Of course, Argon2 would win against iterated AES-NI if its m were much higher (was 16 MiB in this test).

**Comment by LightBit** [comment]
Sorry for just jumping in.

How does Lyra (Lyra 1 not 2) compare to Balloon, Argon2, and Lyra2?
I'm looking for something simple, preferably generic sponge/hash based. I hoped BKDF will be that.

**Comment by solardiz** [comment]
@LightBit I didn't recall (it's been a decade), but I've just skimmed the Lyra paper again. I think Lyra is much more practical compared to Balloon (it uses larger blocks aka longer rows), and is on par with scrypt. It claims (and perhaps actually achieves) some TMTO resistance (unlike scrypt, which is deliberately TMTO-friendly). I expect its actual TMTO resistance may vary by the sponge's internal state size (those mentioned for suggested sponges are somewhat small, in hundreds of bits vs. yescrypt's 12 KiB). This internal state is also what discourages sub-block (aka column) pipelining attacks, where scrypt uses sub-block shuffling instead (so Lyra managed to simplify this out). However, Argon2 designers' "Tradeoff Cryptanalysis of Memory-Hard Functions" claims even Lyra2 is vulnerable: "the time-memory product for Lyra2 can be decreased by the factor of 8." I don't recall the detail, such as whether the Lyra2 designers provided a response and what it was.

Incidentally, I admit I didn't get those pipelining attacks before that paper (it was eye-opening for me, even though @cperciva clearly had thought of them during scrypt design), and I adjusted the final yescrypt design in response. This is where yescrypt block function's rapidly updated internal state appeared. But to my knowledge, those attacks were not of practical help against actual uses of earlier yescrypt drafts for PoW and thus mining on GPUs. Maybe they would be on ASICs.

Lyra's reported defensive speeds when instantiated with a BLAKE2-based sponge are on par with scrypt. It uses data-dependent addressing (which makes me wonder why they suggest higher than minimum T - perhaps again a chase for academic levels of TMTO resistance beyond practical reason, same as we see in recommended parameters for Balloon and Argon2). Lyra2 is a safer bet, given it's a newer yet already mature revision, and is PHC's special recognition. It's also only moderately more complex. It further improves TMTO resistance while achieving similar speeds, and integrates BlaMka (for multiplication latency hardening, which is also implemented in Argon2 and yescrypt).

So if you really want a single-primitive scheme and don't mind the sub-block pipelining risk (which apparently no one exploited in practice so far), Lyra2 isn't bad, and even the original Lyra looks adequate too. However, what single primitive would that be? With the recommended ones based on pieces from BLAKE2, you'd have the same standards compliance issues as with Argon2.

In (ye)scrypt, at least you have entropy bypass via proven and standards-compliant HMAC and SHA-256, so you can prove that basic cryptographic properties are inherited from those primitives (even if the heavy processing completely fails, like returns all zeroes). For scrypt, this is the case for the password input to hash output. For yescrypt, it's also for the salt input.

I see you've already implemented scrypt before, so you actually know it's not all that complex. Why not just continue using it if you don't want anything more complex and are fine with data-dependent indexing? https://github.com/LightBit/libkripto/blob/master/lib/scrypt.c

**Comment by LightBit** [comment]
@solardiz Thanks. Will check this paper and Lyra2 revisions.
Its not really hard to implement with reference implementation available, but I prefer simpler designs that are easy to understand (idealy so simple you can implement it from head).

**Comment by samuel-lucas6** [comment]
Sorry for the delayed response again.

> For Argon2 as we currently have it in JtR, this point is in the hundreds of megabytes, which is rather high.
> ... although some other attacker would choose a bcrypt-friendlier GPU instead (perhaps recent NVIDIA).

Thank you @solardiz, I see. How equivalent is increasing the iterations with lower memory?

> I couldn't quickly find where they say "the Argon2 round function can be made more sequential".

A summary is on page 8, and the more detailed version is Section 7. They modify the round function by adding some data-dependencies using XOR. According to the authors, it roughly preserves defender performance whilst slowing down the attacker.

> To me, trapdoor memory-hard functions sound like they're for some different use cases, and I wonder what those would be.

From that conversation, my understanding was the user keeps a secret key, whereas the server stores a public key (kind of like passkeys). The user can compute the output using little memory, whereas the server has to use lots of memory. That way, if the server is compromised, an attacker has to use lots of memory as well. Although this seems kind of backwards in that servers may be more limited than clients since they have to handle multiple clients.

My thinking for why it's worse than a pepper/encrypting hashes is that those approaches prevent password cracking completely unless the secret is known. And even if the secret is known, you still have the memory-hard function to slow down cracking. With trapdoor MHFs, it sounds like a compromise of the secret ruins the defence in terms of memory. They're also more complicated to understand and implement, so it doesn't sound like they'll take off. PAKEs don't seem to have been that successful, and now we have passkeys assuming the UX gets sorted.

> As to pepper, to me it's clear (and was at the time of PHC) that we need hash encryption instead of pepper (which is why yescrypt only has hash encryption). It's as good as pepper, but it also allows for key change (if potentially compromised). Doing it without dependency on an extra primitive is a bit tricky. yescrypt has a custom Feistel cipher based on SHA-256. It's not very complex, but I understand that when you optimize for simplicity, you implement a pepper instead.

I agree for password hashing. However, I think a pepper makes sense for key derivation in that it's one-way like the rest of key derivation.

The nice thing about a design like Finnick is because it's based on an encryption algorithm, you can use the underlying function for that purpose.

> Looks like you misunderstood me. I didn't comment on Argon2's explicit/configurable parallelism at all. I commented on the lower-level parallelism hard-coded in its block function

Ah, sorry. I think I see what you mean. You've made a lot of nice presentations I need to take a look at.

> Sure AES can be fast on GPUs, but AES-NI makes it faster than other primitives on CPUs, so the GPU advantage is reduced. As to this specific paper, we actually looked into it and made code changes in JtR accordingly:

That makes sense. I'm surprised there hasn't been more use of AES, although I suppose hardware support has increased since the PHC. Do you think lack of support is really a problem now? I would think it's mainly an issue for embedded devices, which aren't going to be using a memory-hard function anyway.

> I'm looking for something simple, preferably generic sponge/hash based. I hoped BKDF will be that.

@LightBit so did I. The issue is standard hash functions don't seem ideal for this purpose. You want something with a larger output and a function that's suboptimal for hardware (aka not SHA-3, which is obviously the main sponge-based family and going to become more popular with PQC). And if the password hashing algorithm is significantly slower than alternatives, it shouldn't be used from a security standpoint because you're weakening yourself.

I like simplicity a lot too, but it may be the best password hashing algorithms need to be a bit more complex, which definitely isn't the case with some other types of algorithm. Having said that, perhaps cache-hard algorithms don't need to be as complicated. There needs to be more documentation on them.

> perhaps again a chase for academic levels of TMTO resistance beyond practical reason, same as we see in recommended parameters for Balloon and Argon2

Which parameters are you talking about? Do you mean like increasing the delta in Balloon, high memory/iterations, or something else? The Argon2 RFC has very high memory recommendations that don't sound practical.

**Comment by solardiz** [comment]
> How equivalent is increasing the iterations with lower memory?

I'm not sure I get this question. Do you mean when memory usage is so low anyway that we haven't yet reached the point where increases of it slow down GPUs more than CPUs? If so, there's in fact a range of parameters where increasing iterations is only slightly less effective against GPUs than increasing memory, as either parameter has a close to linear effect. However, increasing memory _is_ slightly more effective - and more than slightly against ASICs - even at these low parameters.

Where I mentioned that a Monero ASIC at 2 MiB was ~100x more power-efficient than CPUs and GPUs, you need to consider that the factor of ~100x is still way lower than it'd have been for PBKDF2 (~1k vs. GPUs and ~10k vs. CPUs, as we see from Bitcoin mining).

> > I couldn't quickly find where they say "the Argon2 round function can be made more sequential".
> 
> A summary is on page 8, and the more detailed version is Section 7. They modify the round function by adding some data-dependencies using XOR. According to the authors, it roughly preserves defender performance whilst slowing down the attacker.

Oh, sure. I wish they clearly marked the changes in the referenced GitHub repo, such as made them a separate commit.

But anyway, yes, this was a known property of Argon2, and yes it could be made more sequential without decrease in throughput on a server, see e.g. last paragraph in my comment below comparing memory bandwidth usage with yescrypt, which has 6x more sequential multiplication latency hardening:

https://github.com/P-H-C/phc-winner-argon2/issues/231

Further, in there I show that even higher defensive speeds are possible despite of the more sequential computation if the block size is increased a bit (it's fixed at 1 KiB in Argon2, but is configurable in (ye)scrypt, so for yescrypt I currently recommend 4 KiB).

Also see the `COMPARISON` in my upstream yescrypt tree and on its homepage. Which, by the way, I had asked others on the PHC list to comment when the list was still operational (even though PHC was over), but no one did, so I assume I got everything right.

Unfortunately, many in PHC were too focused on single thread latency under low load, whereas what limits the parameter values on servers is throughput under load spike or simulated DoS-like conditions. When I was consulting on yescrypt deployments, the task was always to achieve this-many requests per second per server at full (aka abnormally high) load. Latency also mattered, but again under those same max concurrent load conditions (latency under low load is then obviously acceptable and more stable anyway).

On trapdoors:

> From that conversation, my understanding was the user keeps a secret key, whereas the server stores a public key (kind of like passkeys). The user can compute the output using little memory, whereas the server has to use lots of memory.

Oh, then my _guess_ is it may be intended as a way to authenticate quickly while the secret key is available, but also be able to authenticate (slower) even if the secret key is lost. This makes some sense then. However, in a resource exhaustion DoS attack the slow, server-heavy path would be triggered all the time. So this accepts that this slow/recovery path may have worse availability (the server may have to disallow it first while under attack).

> My thinking for why it's worse than a pepper/encrypting hashes is that those approaches prevent password cracking completely unless the secret is known.

Yes, but maybe they have different goals - see above. Also, maybe they can co-exist with hash encryption on server.

> I think a pepper makes sense for key derivation in that it's one-way like the rest of key derivation.

With key derivation, where would you store the pepper such that it'd not always leak when the salt does? Well, maybe smartcard/TPM could have it, but salts be part of encrypted files.

As to it being one-way, what does this buy us? I think almost nothing. If we compare against encryption of the KDF output, which can be reversed, it's just a tiny fraction of the total running time of the KDF. Sure a password cracker could reverse it at startup, but with a heavy enough KDF the overall effect of that reversal would be negligible. So I think it's preferable not to special-case the KDF use case, and only have hash/KDF encryption rather than pepper.

> I'm surprised there hasn't been more use of AES, although I suppose hardware support has increased since the PHC. Do you think lack of support is really a problem now?

It's less of a problem than it was at the time of PHC. However, I think the biggest issue is that when using AES, we have to supply a software AES implementation in the tree anyway, even if it would be very rarely used. Also separate code for x86 and ARM, or else ARM would fallback to software and not only be slower but also have extra side-channel leaks. So the source tree is complicated significantly. Unless it's a higher-level implementation that uses a library that has AES anyway. Further, if you use not the full AES, but pieces from it (like individual rounds), then you generally can't rely on a library providing those (have to carry your own).

> perhaps cache-hard algorithms don't need to be as complicated.

Unfortunately, cache-hard-only algorithms only defend against GPUs, somewhat against FPGAs (if targeting L2+ cache on CPUs, not L1 like bcrypt did), and do quite little against ASICs. Also, no attempt to make them data-independent so far.

The cache-hardness in yescrypt adds little complexity on top of what yescrypt needed anyway. Maybe even none. The internal state is also needed against sub-block pipelining anyway.

> > perhaps again a chase for academic levels of TMTO resistance beyond practical reason, same as we see in recommended parameters for Balloon and Argon2
> 
> Which parameters are you talking about?

The "time complexity" or t_cost or t or T or iterations, whatever it's called in a given algorithm. The academic preference was to set it high enough that TMTO resistance is obviously prohibitively high. However, this fails to consider that setting it lower enables much higher total defensive memory usage in same total running time, so even moderately effective TMTO would still result in higher memory usage since the starting point would have been so much higher. Like "one half of 30 MB" (lowest t_cost, 30 MB total, lowered by a factor of 2 by TMTO in an attack) vs. "all of 10 MB" (t_cost=3 defeating that TMTO "fully", but limiting defender's usage to 10 MB). For practical purposes, from a defender's perspective, 15 MB post-TMTO is better than 10 MB without TMTO.

And yes, in Balloon also delta, but if instantiated with a slow primitive like SHA-512 this doesn't hurt - doesn't slow it down much _more_ on top of the slow primitive. So it's actually reasonable to have high delta on top of that. It wouldn't be so reasonable with a properly fast large-output primitive.

> The Argon2 RFC has very high memory recommendations that don't sound practical.

The Argon2 RFC's section "[4. ](https://www.rfc-editor.org/rfc/rfc9106.html#section-4)[Parameter Choice](https://www.rfc-editor.org/rfc/rfc9106.html#name-parameter-choice)" looks detached from reality. Not only are those target runtimes much higher than what's typically required for authentication, but they also suggest usage of 2+ lanes even for cryptocurrency mining and authentication. Somehow I see the RFC lists me among those "who helped in preparing and reviewing this document", which I appreciate, but I don't recall reviewing these parameters. On the bright side, they do appear to favor t=1. :-)

**Comment by solardiz** [comment]
> in Balloon also delta, but if instantiated with a slow primitive like SHA-512 this doesn't hurt - doesn't slow it down much _more_ on top of the slow primitive. So it's actually reasonable to have high delta on top of that.

Sorry, I misrecalled how Balloon works. Higher delta increases the number of hash computations (the way the algorithm is given in the paper) or at least the amount of data to hash (the way you have it in the BKDF spec), not only the number of memory reads (as I had wrongly recalled), so it makes e.g. SHA-512 a lot slower at producing output. So the same concerns apply as to higher t_cost. This probably _does_ result in a lot lower s_cost, and any changes in TMTO resistance should be considered with that in mind, not on their own.

**Comment by solardiz** [comment]
@samuel-lucas6 I found the discussion you had in https://crypto.stackexchange.com/questions/114049/whats-the-use-of-the-delta-parameter-in-balloons-hash and also took a fresh look at the Balloon paper. In that StackExchange discussion, the comment by n-l-i is helpful, but it somehow assumes t_cost does not contribute to TMTO resistance at all, whereas I think it also does. The attacks on Argon2i (specific to data-independent indexing) in the Balloon paper give different TMTO factors depending on t_cost, namely 1/4 for lowest t_cost (single-pass) vs. 1/e for high t_cost (a very impressive result). I see how the 1/4 vs. 1/e difference may be deemed insignificant compared to the greater effect delta has. So maybe in Balloon / BKDF delta=3 is actually optimal. You also don't want lower delta because you'd deviate from Balloon, which would partially defeat the purpose of this project.

**Comment by solardiz** [comment]
I forgot that Argon2i has been updated to address those attacks. They were for version 1.2.1, whereas in version 1.3 that is in use now "The blocks are XORed with, not overwritten in the second pass and later". I see that Balloon also processes the block being written to as input, which should have similar effect. With this, I expect that increases in t_cost should have significant effect on TMTO resistance of both Argon2i and Balloon/BKDF. However, this is not a reason to use delta lower than 3 and compensate for that with higher t_cost. Either high delta or high t_cost result in lower s_cost anyway. So delta=3, t_cost=1, and s_cost as high as affordable looks reasonable for Balloon.

**Comment by samuel-lucas6** [comment]
> However, increasing memory is slightly more effective - and more than slightly against ASICs - even at these low parameters.

Yes, I've heard increasing memory is better than increasing the iterations. That's what I wanted to confirm.

> But anyway, yes, this was a known property of Argon2, and yes it could be made more sequential without decrease in throughput on a server

Ah right. I assumed it was new to that paper.

> Further, in there I show that even higher defensive speeds are possible despite of the more sequential computation if the block size is increased a bit (it's fixed at 1 KiB in Argon2, but is configurable in (ye)scrypt, so for yescrypt I currently recommend 4 KiB).

Getting the right block size was one thing I wasn't sure about with Finnick.

> Unfortunately, many in PHC were too focused on single thread latency under low load, whereas what limits the parameter values on servers is throughput under load spike or simulated DoS-like conditions.

That's interesting. Part of the issue is you probably need to have worked on a system like that to find that out.

> Oh, then my guess is it may be intended as a way to authenticate quickly while the secret key is available, but also be able to authenticate (slower) even if the secret key is lost.
>
> Yes, but maybe they have different goals - see above. Also, maybe they can co-exist with hash encryption on server.

That's a good point.

> With key derivation, where would you store the pepper such that it'd not always leak when the salt does? Well, maybe smartcard/TPM could have it, but salts be part of encrypted files.

In a file encryption scenario, it can be done with a YubiKey (challenge-response) or a keyfile (e.g., stored on a USB drive). 1Password also uses a pepper that's somehow stored on client devices.

> As to it being one-way, what does this buy us? I think almost nothing.

Yeah, I agree. It was just an observation. I don't think it makes much of a difference.

> It's less of a problem than it was at the time of PHC. However, I think the biggest issue is that when using AES, we have to supply a software AES implementation in the tree anyway, even if it would be very rarely used. Also separate code for x86 and ARM, or else ARM would fallback to software and not only be slower but also have extra side-channel leaks. So the source tree is complicated significantly.

That's fair. The newer constructions use the AES round function, so that helps a bit in terms of space.

> Unfortunately, cache-hard-only algorithms only defend against GPUs, somewhat against FPGAs (if targeting L2+ cache on CPUs, not L1 like bcrypt did), and do quite little against ASICs. Also, no attempt to make them data-independent so far.
>
> The cache-hardness in yescrypt adds little complexity on top of what yescrypt needed anyway. Maybe even none. The internal state is also needed against sub-block pipelining anyway.

Got it, so a combined approach is probably best. Do you know why [Argon2ds](https://www.password-hashing.net/argon2-specs.pdf) never took off?

Also, is there any good reading material besides your [yescrypt specification](https://www.password-hashing.net/submissions/specs/yescrypt-v2.pdf) for understanding how to implement cache-hardness? Argon2ds isn't really explained, and [Pufferfish2](https://github.com/epixoip/pufferfish)/[bscrypt](https://github.com/Sc00bz/bscrypt) have no specification.

> The academic preference was to set it high enough that TMTO resistance is obviously prohibitively high. However, this fails to consider that setting it lower enables much higher total defensive memory usage in same total running time

Oh yeah, I remember reading a paper recommending more iterations.

It would be nice to have as few parameters as possible, like PBKDF2/bcrypt, but I guess that's not possible without leaving some extra security on the table.

> I found the discussion you had in https://crypto.stackexchange.com/questions/114049/whats-the-use-of-the-delta-parameter-in-balloons-hash

That was a good question but a confusing discussion. It would be nice if more authors anticipated these types of questions and answered them in the paper/specification.

> So maybe in Balloon / BKDF delta=3 is actually optimal.
>
> So delta=3, t_cost=1, and s_cost as high as affordable looks reasonable for Balloon.

That's good. Thanks for that recommendation.

**Comment by solardiz** [comment]
> Getting the right block size was one thing I wasn't sure about with Finnick.

Yes, it's tricky. You need to optimize defensive (vs. offensive) speeds for direct computation, but also consider that the way you process sub-blocks in a block may be vulnerable to pipelining. So in yescrypt I wouldn't recommend blocks larger than pwxform's internal state (currently 12 KiB). Luckily, defensive speeds are good at 4 KiB, so we don't reach that level.

Also consider that by a larger block, you may be taking more of your defensive cache space away from possible more cache-hardness. OTOH, without deliberate cache-hardness in the algorithm anyway, using a larger block size may provide a tiny bit of poor man's cache-hardness-alike (CPU will have the previous block in cache, but GPU at high concurrency might have to re-fetch it piece-by-piece, or otherwise have its concurrency more limited) - but this is a risky bet.

> Got it, so a combined approach is probably best. Do you know why [Argon2ds](https://www.password-hashing.net/argon2-specs.pdf) never took off?

I don't know exactly why, but I suspect it's because:
1. PHC chose Argon2 as the winner largely for what the judges saw as simplicity (although to me it doesn't actually look simpler than scrypt). Any extras would go against that.
2. 2id and 2ds were late optional mode additions. Since 2id did take off, this means that this alone isn't enough of a reason why 2ds did not.
3. The reference PHC tree at https://github.com/P-H-C/phc-winner-argon2 added 2id, but not 2ds. So it was this tree maintainers' decision to support/promote 2id and not 2ds. They understandably didn't want too many variants.
4. That decision may have been based on already existing demand/criticism. As I recall, 2id was in fact requested by some on the PHC list.
5. No 2ids mode was defined. If someone wanted defense against cache timing, they had to choose 2i or 2id. So 2id it was.
6. Out of most-active participants on the PHC list, the properties of 2ds in a candidate algorithm (that is, sequential memory-hard and also [what later became known as] cache-hard) were requested by Bill Cox @waywardgeek and me. But both of us had our own candidates, and yescrypt got special recognition. So Argon2ds would be kind of redundant with yescrypt.
7. Password crackers were late to add Argon2 support/attacks. Our integration of Argon2 OpenCL code into JtR is late 2023. We don't yet have yescrypt 1.0 in OpenCL, at all. So we're just now starting to truly show what these differences mean in practice.

BTW, 2ds adds not only cache-hardness, but also latency hardening - both on par with yescrypt 0.5, which is not a coincidence. But differently from yescrypt, IIRC it uses the approach that TwoCats also chose - the latency-hardened chain is scalar. So while the SIMD units compute the 1 KiB block in parallel, scalar units have to go through the 96 rounds sequentially. This means that there's a lot less computation total (sequential times parallel) in that chain (compared to yescrypt's way of doing cache-hardness and latency hardening as part of its main block processing on SIMD units). But most of the desired effect should be achieved anyway.

> Also, is there any good reading material besides your [yescrypt specification](https://www.password-hashing.net/submissions/specs/yescrypt-v2.pdf) for understanding how to implement cache-hardness?

Unfortunately, no. The best reading material may be PHC list postings by Bill Cox and me. ;-)

The way I approached it, I always kept two kinds of attack in mind - with S-boxes in local vs. global memory. In bcrypt, the lookups are so tiny and frequent that only local memory attacks are practical. (You can do global memory indeed, but it's just slower. You could maybe do both at once for higher cumulative throughput, which we briefly attempted, but no luck so far.)

As you start increasing the S-boxes to target L2+ rather than L1 cache, and have to make lookups less frequent, you on one hand make local memory attacks even slower, but on the other hand make global memory attacks more practical. In other words, target too high a cache level and size, and you're no longer local memory hard, so this stops being a separate defense.

The width of lookups also matters, but there's no drawback from wider lookups for the defender as long as this width increase does not make them less frequent nor less numerous nor more parallel. This is why bcrypt 32-bit vs. yescrypt 128-bit does not make yescrypt weaker in typical builds.

During PHC, I felt that being obviously/provably (by specific metrics) no worse (or at least not much worse) than bcrypt is the important goal. So all of yescrypt, TwoCats, and Argon2ds target L1 cache by default. Later yescrypt 0.5 mining results on specific GPUs (circa 2018) show that this was still pretty effective, but there was room for improvement, at least short-term to satisfy the needs of some altcoins. So while yescrypt 1.0 uses 12 KiB, yespower 1.0 went to 96 KiB at twice fewer pwxform rounds (so twice less frequent lookups against mostly L2 cache, increasing vulnerability to global memory attacks but not enough to make them most practical). I conservatively kept it at 128-bit lookup width (to be able to reason based on existing yescrypt 0.5 mining speeds), although going for 512-bit is also reasonable when targeting L2+ cache (but would be harder to predict the full effect of). All of these choices were based on benchmarks on hardware most relevant at the time.

The way yescrypt/yespower pwxform is defined, all of these things are tunable (currently compile-time). You may want to see these slides: https://www.openwall.com/presentations/BSidesLjubljana2017-Yescrypt-Large-scale-Password-Hashing/slide-044.html

With today's L1d commonly increased from 32 to 48 KiB, I'd probably choose at least 24 KiB for yescrypt, but for now I feel this is not enough of a reason to re-tune and introduce a variation.

> It would be nice to have as few parameters as possible, like PBKDF2/bcrypt, but I guess that's not possible without leaving some extra security on the table.

Right, but it's reasonable to have a wrapper that would map one parameter onto the rest, with the full derived parameter set recorded along with the hashes. This is done in libxcrypt for yescrypt:

https://github.com/besser82/libxcrypt/blob/e34e0634a7192383dbb4445b6d7568e36baef675/lib/crypt-yescrypt.c#L136

**Comment by waywardgeek** [comment]
Wow, this is a blast from the  past!  It has been a decade since the Password Hashing Competition.  I just build a distributed oblivious pseudo-random oracle service for encrypting with a pin/password.  If you feel inspired, feel free to checkout openadp on github.

**Comment by samuel-lucas6** [comment]
Thanks for answering all my questions and commenting here in the first place @solardiz. I did read this last weekend, but I wasn't at my desk.

I think my next steps need to be looking at more PHC candidates and implementing Finnick properly to see how much faster that is than BKDF/Balloon, but I don't know when I'm going to get around to doing that due to other commitments. I can share what the performance is like here if you're interested, although I'm certain there are better designs. Hopefully the research will pick up again at some point and whoever writes the next cryptography book will include a chapter on password hashing.
